docs/operations/batch-processing.md

```diff
@@ -0,0 +1,132 @@
+# Batch Processing
+
+In order to answer queries, JanusGraph has to perform queries
+against the storage backend.
+In general, there are two ways of doing this:
+
+- Once data from the backend is needed, execute a backend
+  query and continue with the result.
+- Maintain a list of what data is needed.
+  Once the list reaches a certain size, execute a single backend
```

Comment:
I didn't even know that's how the CQL backend handles MultiQueries, thanks for clarifying! But now that I know that, I'm even more impressed that MultiQueries actually make such a differernce. When I first tested the old implementation, I was able to reach performance improvements of more than 300x in extreme cases just by enabling `query.batch`. I would never have thought that's possible only by firing queries in parallel!
But now that I know it, I will try to find another name for it ;)

Reply:
300x sounds too good to be true unless your server has 150 cores :)

FYI the relevant code is in BackendTransaction::edgeStoreMultiQuery, which uses a thread pool with size  = numOfCores * 2.

Reply:
The sheer number of threads is not the only thing that makes MultiQueries faster, it should also lead to much better code locality which means fewer instruction cache misses. I tested it on a server with 24 cores.

Now the `numOfCores * 2` is very interestring and I think we should have a closer look at that as soon as possible for two reasons:
- Since most of what the scheduled threads will do is stalling, I suspect the capabilities of `n` cores should be way beyond `2n` backend queries.
- I don't know how the JVM handles `Runtime.getRuntime().availableProcessors()` when you limit the available CPUs for a docker container, but we should definitely be aware that many clients are probably running JG in such a configuration (including us).

In addition, I don't really know what happens if JanusGraph is concurrently executing multiple user queries and how that influences the number of parallel backend query threads.

Reply:
I don't know if it could lead to fewer instruction cache misses, but one thing is: multi-threading brings about context switches cost, which should be dominant over code locality. Thus, the performance boost should be less than numOfCores * 2.

> Since most of what the scheduled threads will do is stalling, I suspect the capabilities of n cores should be way beyond 2n backend queries.

I agree. FYI there is an issue for this: https://github.com/JanusGraph/janusgraph/issues/2473 @haocornell-ms once expressed interest to work on that on gitter channel but I am not sure how it's going.

Reply:
I actually suspect the performance boost scale is bound by 10, due to the corePoolSize of ThreadPoolExecutor in CQLStoreManager: https://github.com/JanusGraph/janusgraph/blob/fff871c6b033c1f01724353a89bab3dcb16efd89/janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLStoreManager.java#L183-L191

Assuming `db-cache` is disabled, then at a time, there are at maximum 10 threads executing CQL queries. Thus the upper bound shall be 10x.